### PR TITLE
Add a generic RO-Crate Parser

### PR DESCRIPTION
**Pull request by JLoveUOA**
Added a profile that parses an RO-Crate (based on known fields from schema.org) into indigestible data classes.

consisting of:
- samples/ro_crate_ingestion_example.py
an example ingestion script based off the Biru ingestion example

-  src/profiles/ro_crate/ro_crate_parser.py  
Parser for reading RO-Crates into indigestible data-classes.
    - takes the input of the directory containing their ro-crate-metadata.json and a schema file (to be updated to an RO-Crate profile) for mapping the schema.org field names to MyTardis names.
    - Translates RO-Crate Projects, Data-Catalogs, Datasets and Files into MT Projects, Experiments, Datasets and Datafiles by reading their JSON-LD dictionary into an appropriate ingestible data-class, translating fields according to provided schema and de-referencing linked fields such as people and metadata.
    - Reads through child directories of Dataclasses for files not explicitly listed in the RO-Crate JSON and constructs Datafile Objects
    - metadata is determined by MyTardis metadata entities for each metadata element. MyTardis metadata entities in an RO-Crate follow the schema of:
        ```json         
        "@id": (unique string) <RO-Crate Entity ID>,
        "@type": (string)"MyTardis-Metadata_field",
        "name": (string) <MyTardis metadata field name>,
        "value": (given by mt-type) <value>,
        "mt-type": (string) <MyTardis Metadata Type>,
        "sensitive":(Boolean)
        ```

- src/profiles/ro_crate/crate_to_tardis_mapper.py : A helper class for mapping from Schema.org fields to their corresponding MyTardis fields.

- src/profiles/ro_crate/test_mapping.json : A json file used by the above class to determine mappings

- Added the RO-Crate library (which added a LOT to poetry.lock)

TODO:

- Incorporate RO-Crate parsing into new profiles and ingestion system
- Write tests for RO-Crate parsing
- Create schema for custom RO-Crate feilds (those that do not appear in schema.org)
- Replace mapping schema with an RO-Crate profile

### CODE REVIEW

File: src/profiles/ro_crate/ro_crate_parser.py [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+
+        dataset_dict["experiments"] = get_norm_value(
+            crate_dataset.as_jsonld(), "includedInDataCatalog"
+        )
+        if len(dataset_dict["experiments"]) == 0:
+            raise ValueError(
+                "Dataset "
+                + str(crate_dataset.id)
+                + " not included in a DataCatalog, cannot link this dataset to experiments"
+            )
```
Comment (uoa-noel) [inline src/profiles/ro_crate/ro_crate_parser.py:184]: As this relates to a particular Experiment, it's better to either use `logger.error()`, or catch the error and print out a error logging message, so the script doesn't stop running on one invalid Experiment.

**Review by uoa-noel — CHANGES_REQUESTED** [review]
Looking great James, I just see two minor things in relation to logging/error handling then it's ready to be merged. There are a few TODOs you've outlined, I guess they will be part of a separate PR.

File: src/profiles/ro_crate/ro_crate_parser.py [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+        if rocrate_entity:
+            rocrate_dict: dict[str, Any] = {
+                (self.mapper.get_mt_field(MyTardisObject.DATAFILE, key) or key): value
+                for key, value in rocrate_entity.as_jsonld().items()
+            }
+            datafile_dict.update(rocrate_dict)
+            if datafile_meta := datafile_dict.get("metadata"):
+                datafile_dict["metadata"] = self._read_metadata([datafile_meta])
+        datafile_dict["dataset"] = parent_dataset_description
+        print(datafile_dict["filename"], "my parent is,", datafile_dict["dataset"])
```
[the remark this answers is not in the document]
Comment (JLoveUOA) [reply src/profiles/ro_crate/ro_crate_parser.py]: whoops just completely missed removing that one on my final debug cleanup :smile:

File: src/profiles/ro_crate/ro_crate_parser.py (hunk 1 above)
Comment (JLoveUOA) [reply src/profiles/ro_crate/ro_crate_parser.py]: changed to a logger.error()

**Review by uoa-noel — APPROVED** [review]
LGTM :+1:

**Pull request merged**